src/Elastic.Apm.AspNetCore/Extensions/TransactionExtentions.cs

```diff
@@ -1,36 +1,52 @@
 using System.Net.Mime;
-using Elastic.Apm.Api;
 using Elastic.Apm.Config;
 using Elastic.Apm.Logging;
+using Elastic.Apm.Model;
 using Microsoft.AspNetCore.Http;
 
 namespace Elastic.Apm.AspNetCore.Extensions
 {
-	public static class TransactionExtensions
+	internal static class TransactionExtensions
 	{
 		/// <summary>
-		/// Collects the Request body (and possibly additional information in the future)
-		/// in the transaction
+		/// Collects the Request body in the transaction
 		/// </summary>
 		/// <param name="transaction">Transaction object</param>
-		/// <param name="httpContext">Current http context</param>
+		/// <param name="isForError">Is request body being captured for error (otherwise it's for transaction)</param>
+		/// <param name="request">Current http request</param>
 		/// <param name="logger">Logger object</param>
-		public static void CollectRequestInfo(this ITransaction transaction, HttpContext httpContext, IConfigurationReader configurationReader,
-			IApmLogger logger
-		)
+		internal static void CollectRequestBody(this Transaction transaction, bool isForError, HttpRequest request, IApmLogger logger)
 		{
-			var body = Consts.BodyRedacted; // According to the documentation - the default value of 'body' is '[Redacted]'
+			if (!transaction.IsSampled) return;
 
-			// We need to parse the content type and check it's not null and is of valid value 
-			if (!string.IsNullOrEmpty(httpContext?.Request?.ContentType))
-			{
-				var contentType = new ContentType(httpContext.Request.ContentType);
+			// Is request body already captured?
+			if (transaction.IsContextCreated
+				&& transaction.Context.Request.Body != null
+				&& !ReferenceEquals(transaction.Context.Request.Body, Apm.Consts.Redacted)) return;
 
-				//Request must not be null and the content type must be matched with the 'captureBodyContentTypes' configured
-				if (httpContext?.Request != null && configurationReader.CaptureBodyContentTypes.ContainsLike(contentType.MediaType))
-					body = httpContext.Request.ExtractRequestBody(logger);
-				transaction.Context.Request.Body = string.IsNullOrEmpty(body) ? Consts.BodyRedacted : body;
-			}
+			string body = null;
+			if (IsCaptureRequestBodyEnabled(transaction, isForError) && IsCaptureRequestBodyEnabledForContentType(transaction, request))
+				body = request.ExtractRequestBody(logger);
+			// According to the documentation - the default value of 'body' is '[Redacted]'
+			transaction.Context.Request.Body = body ?? Apm.Consts.Redacted;
+		}
+
+		private static bool IsCaptureRequestBodyEnabled(Transaction transaction, bool isForError) =>
+			transaction.ConfigSnapshot.CaptureBody.Equals(ConfigConsts.SupportedValues.CaptureBodyAll)
```

Comment:
At least to me `IConfigurationReaderExtentions` with `ShouldExtractRequestBodyOnError` and `ShouldExtractRequestBodyOnTransactions` was easier to follow. - no action needed on this, just saying.